docs/enhancements/cluster_pool_hot_spares.md

```diff
@@ -0,0 +1,171 @@
+# Maintaining Running Clusters in a ClusterPool
+
+[HIVE-1576](https://issues.redhat.com/browse/HIVE-1576)
+
+- [Maintaining Running Clusters in a ClusterPool](#maintaining-running-clusters-in-a-clusterpool)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [User Story](#user-story)
+  - [Proposal](#proposal)
+    - [`spec.runningCount`](#specrunningcount)
+    - [Decouple PowerState on Creation](#decouple-powerstate-on-creation)
+    - [ClusterClaim Assignment by `creationTimestamp`](#clusterclaim-assignment-by-creationtimestamp)
+    - [Hibernation Controller Must Account For `hibernateAfter` in Pool Clusters](#hibernation-controller-must-account-for-hibernateafter-in-pool-clusters)
+    - [A Note About Tuning](#a-note-about-tuning)
+  - [Alternatives](#alternatives)
+    - [Don't Hibernate New ClusterDeployments](#dont-hibernate-new-clusterdeployments)
+    - [`runningPercent`](#runningpercent)
+
+## Summary
+
+Support configuring ClusterPools to maintain some number of clusters (`0 <= N <= spec.size`) in Running state.
+
+## Motivation
+
+Today ClusterPools immediately hibernate new clusters.
+This means a ClusterClaim is guaranteed to suffer at least Resuming time before the cluster is available for use.
+While this is still faster than having to create a cluster from scratch, it is a potentially-unnecessary delay.
+As an example, CI runs could be reduced by ~5m if a ClusterClaim were able to grab a Running cluster rather than having to Resume a Hibernating one.
+
+### Goals
+
+Ability to tune a ClusterPool such that a ClusterClaim gets a Running cluster immediately in the common case.
+
+(Stretch goal) Solving world peace.
+
+### User Story
+
+As a consumer of a ([properly tuned](#a-note-about-tuning)) ClusterPool,
+
+I want ClusterClaim to give me a Running cluster immediately in the common case,
+
+So that I can get my work done more quickly.
+
+(Ideally without sacrificing the cost savings of having some clusters in the pool Hibernating.)
+
+## Proposal
+
+### `spec.runningCount`
+
+Add a field to ClusterPool indicating a portion of the pool to maintain in Running state.
+`spec.runningCount` is an integer absolute number of Running clusters to strive for.
+It is constrained to `0 <= runningCount <= Size`.
+We could enforce this by webhook; or we could just update the spec accordingly (explicitly set `runningCount = min(runningCount, size)`).
+
+If unspecified, the default behavior is to hibernate all CDs in the pool, as today.
+This corresponds to `runningCount=0`, conveniently the golang nil value of an int.
+
+### Decouple PowerState on Creation
+
+Today we create all new ClusterDeployments with `powerState=Hibernating` and update that to `Running` at claim time.
+With this feature, we will create with `powerState` unset.
+Then in a separate code path:
+- Sort unassigned (installing + ready) CDs by `creationTimestamp`
+- Update the oldest `runningCount` CDs with `powerState=Running`
```

Comment:
Will this conflict with the stale cluster deletion, which is also based on oldest?

Reply:
Yes (which wasn't a thing when I wrote this :) )

Today we don't take staleness into account when assigning: we'll assign the oldest installed cluster even if it's stale; but then that takes that cluster out of consideration for deletion due to staleness.

If we don't treat stale clusters specially:
- Kick the oldest cluster, which happens to be stale, to running
- Delete the oldest stale cluster, which is ^ that one
- Replace that cluster. While it's installing...
- Kick the oldest cluster, which happens to be stale, to running

We won't delete another stale cluster until the replacement is done installing. As long as a claim comes in before then, we'll assign the now-running stale cluster that was the second-oldest when we started. If not, we'll repeat the above sequence. It's not really thrashy since it's only happening at worst on a ~40m interval.

Is this something you feel we need to solve?

Reply:
I think the important thing to note here is that we're preserving all of these rules separately (because trying to coordinate them would be insanely complicated):
- Claiming gets the oldest ready cluster -- FIFO
- runningCount hits the oldest ready clusters -- again for FIFO purposes so a given :arrow_up: claim is most likely to get a running cluster
- Delete the oldest stale cluster first -- again for FIFO purposes.